torch_nets/models.py

```diff
@@ -35,55 +49,97 @@ def __init__(
+        features_mean: Any = None,
+        features_std: Any = None,
+        outputs_mean: Any = None,
+        outputs_std: Any = None,
+        output_groups: Any = None,
     ):
-        """Build ``ANN``."""
-        super().__init__()
-        self.linear1 = Linear(n_in, neurons)
-        self.linear2 = Linear(neurons, neurons)
-        self.linear3 = Linear(neurons, neurons)
-        self.linear4 = Linear(neurons, neurons)
-        self.linear5 = Linear(neurons, n_out)
-
-        self.lin_drop1 = Dropout(dropout)
-        self.lin_drop2 = Dropout(dropout)
-        self.lin_drop3 = Dropout(dropout)
-        self.lin_drop4 = Dropout(dropout)
-
-    def forward(self, batch):
-        """Pass ``batch`` through the model.
+        """Initialize the ANN model."""
+        dims: List[int] = [n_in] + [neurons] * (n_layers - 1) + [n_out]
+        layers: List[Any] = []
+
+        for i in range(n_layers):
+            layers.append(nn.Linear(dims[i], dims[i + 1]))
+            if i < n_layers - 1:
+                layers.append(nn.ReLU())
+                layers.append(nn.Dropout(dropout))
+
+        super().__init__(*layers)
+
+        if features_mean is not None:
+            assert features_std is not None
+            assert len(features_mean) == len(features_std)
+            features_mean = torch.tensor(features_mean)
+            features_std = torch.tensor(features_std)
+
+        if outputs_mean is not None:
+            assert outputs_std is not None
+            assert len(outputs_mean) == len(outputs_std)
+            if output_groups is None:
+                outputs_mean = torch.tensor(outputs_mean)
+                outputs_std = torch.tensor(outputs_std)
+            else:
+                assert len(output_groups) == len(outputs_mean)
+                outputs_mean = torch.tensor(
+                    [x for x, g in zip(outputs_mean, output_groups) for _ in range(g)]
+                )
+                outputs_std = torch.tensor(
+                    [x for x, g in zip(outputs_std, output_groups) for _ in range(g)]
+                )
+
+        self.register_buffer("features_mean", features_mean)
+        self.register_buffer("features_std", features_std)
+        self.register_buffer("outputs_mean", outputs_mean)
+        self.register_buffer("outputs_std", outputs_std)
+
+        self.to(torch.device(device))
+
+    def forward(self, input: torch.Tensor):
+        """Pass the input through the model.
 
         Parameters
         ----------
-        batch : Tensor
+        input : Tensor
             A mini-batch of inputs.
 
         Returns
         -------
         Tensor
-            The result of passing ``batch`` through the model.
+            The model output.
 
         """
-        batch = F.relu(self.linear1(batch))
-        batch = self.lin_drop1(batch)
+        if self.features_mean is not None:
+            input = (input - self.features_mean) / self.features_std
+
+        output = super().forward(input)
```

Comment:
I've added a comment to explain it

Reply:
Thanks, but I am not sure this is quite what I had in mind. It can be assumed that `forward` passes input through the NN layers, what I would like clarifying is why you call to `super().forward` as opposed to, say, `self.forward()`.

Reply:
`super().forward` evaluates to the `forward` method of `nn.Sequential`, which passes the input tensor through each of its layers in the sequential order. To call `self.forward()` will lead to an infinite recursion because it is inside the definition of its own `forward` method.

Reply:
I've added the comment

Reply:
This, and the docstring, make it much clearer to others, thank you.